File: nova-android/app/src/test/java/com/nova/agent/PairingTest.kt
package com.nova.agent

import com.nova.agent.net.Pairing
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Bu testler `gateway/test/pairing.test.mjs` ile AYNI vakaları koşar.
 * İki taraf aynı davranmazsa eşleme sessizce bozulur; bu yüzden vakalar
 * bilinçli olarak birebir kopyalanmıştır.
 */
class PairingTest {

    @Test
    fun `alfabede gorsel ikizler yok`() {
        for (ch in "ILOU") {
            assertFalse("$ch alfabede olmamalı", Pairing.ALPHABET.contains(ch))
        }
        assertEquals(32, Pairing.ALPHABET.length)
        assertEquals(32, Pairing.ALPHABET.toSet().size)
    }

    @Test
    fun `normalizeCode gorsel ikizleri duzeltir`() {
        assertEquals("H7K29M4P", Pairing.normalizeCode("h7k2-9m4p"))
        assertEquals("00111234", Pairing.normalizeCode("O0IL1234"))
        assertEquals("H7K29M4P", Pairing.normalizeCode("  H7K2 9M4P  "))
        assertEquals("H7K29M4P", Pairing.normalizeCode("H7K2_9M4P"))
    }

    @Test
    fun `normalizeCode gecersiz girdide kismi kod uydurmaz`() {
        assertEquals("", Pairing.normalizeCode(""))
        assertEquals("", Pairing.normalizeCode(null))
        assertEquals("", Pairing.normalizeCode("H7K2"))
        assertEquals("", Pairing.normalizeCode("H7K29M4PX"))
        assertEquals("", Pairing.normalizeCode("H7K29M4U"))
        assertEquals("", Pairing.normalizeCode("H7K29M4!"))
    }

    @Test
    fun `formatCode insan gozu icin ayirir`() {
        assertEquals("H7K2-9M4P", Pairing.formatCode("h7k29m4p"))
        assertEquals("bozuk", Pairing.formatCode("bozuk"))
    }

    @Test
    fun `baseUrl her zaman v1 ile biter`() {
        assertEquals("http://192.168.1.20:8088/v1", Pairing.baseUrl("192.168.1.20", 8088))
        assertEquals("https://horus.local/v1", Pairing.baseUrl("horus.local", 443, tls = true))
        assertEquals("http://horus.local/v1", Pairing.baseUrl("horus.local", 80))
    }

    @Test
    fun `baseUrl ipv6 adresini koseli paranteze alir`() {
        assertEquals("http://[fe80::1]:8088/v1", Pairing.baseUrl("fe80::1", 8088))
        assertEquals("http://[fe80::1]:8088/v1", Pairing.baseUrl("[fe80::1]", 8088))
    }

    @Test
    fun `baseUrl gecersiz girdide bos doner`() {
        assertEquals("", Pairing.baseUrl("", 8088))
        assertEquals("", Pairing.baseUrl("1.2.3.4", 0))
        assertEquals("", Pairing.baseUrl("1.2.3.4", 70000))
    }

    @Test
    fun `parsePairUri tam uriyi cozer`() {
        val parsed = Pairing.parsePairUri(
            "horus://pair?v=1&code=H7K29M4P&host=192.168.1.20&port=8088&name=SALIH-PC",
        )
        assertNotNull(parsed)
        requireNotNull(parsed)
        assertEquals(1, parsed.version)
        assertEquals("H7K29M4P", parsed.code)
        assertEquals("192.168.1.20", parsed.host)
        assertEquals(8088, parsed.port)
        assertFalse(parsed.tls)
        assertEquals("http://192.168.1.20:8088/v1", parsed.baseUrl)
        assertEquals("SALIH-PC", parsed.name)
    }

    @Test
    fun `parsePairUri kucuk harfli kodu normalize eder`() {
        val parsed = Pairing.parsePairUri("horus://pair?code=h7k2-9m4p&host=1.2.3.4&port=8088")
        assertEquals("H7K29M4P", parsed?.code)
    }

    @Test
    fun `parsePairUri tls bayragini tasir`() {
        val parsed = Pairing.parsePairUri("horus://pair?code=H7K29M4P&host=horus.example&port=443&tls=1")
        assertEquals("https://horus.example/v1", parsed?.baseUrl)
        assertEquals(true, parsed?.tls)
    }

    @Test
    fun `parsePairUri v parametresi yoksa 1 varsayar`() {
        val parsed = Pairing.parsePairUri("horus://pair?code=H7K29M4P&host=1.2.3.4&port=8088")
        assertEquals(1, parsed?.version)
    }

    @Test
    fun `parsePairUri taninmayan girdide null doner`() {
        assertNull(Pairing.parsePairUri(""))
        assertNull(Pairing.parsePairUri(null))
        assertNull(Pairing.parsePairUri("https://example.com"))
        assertNull(Pairing.parsePairUri("horus://other?code=H7K29M4P"))
        assertNull(Pairing.parsePairUri("horus://pair?v=2&code=H7K29M4P&host=1.2.3.4&port=1"))
        assertNull(Pairing.parsePairUri("horus://pair?code=H7K29M4P"))
        assertNull(Pairing.parsePairUri("horus://pair?host=1.2.3.4&port=8088"))
        assertNull(Pairing.parsePairUri("horus://pair?code=H7K29M4P&host=1.2.3.4&port=abc"))
    }

    @Test
    fun `parsePairUri sema buyuk harfle yazilsa da kabul edilir`() {
        assertNotNull(Pairing.parsePairUri("HORUS://PAIR?code=H7K29M4P&host=1.2.3.4&port=8088"))
    }

    @Test
    fun `parsePairUri yuzde kodlamasini cozer`() {
        val parsed = Pairing.parsePairUri("horus://pair?code=H7K29M4P&host=1.2.3.4&port=8088&name=SALIH%20PC")
        assertEquals("SALIH PC", parsed?.name)
    }

    @Test
    fun `parsePairUri isim alanini sinirlar`() {
        val long = "A".repeat(200)
        val parsed = Pairing.parsePairUri("horus://pair?code=H7K29M4P&host=1.2.3.4&port=8088&name=$long")
        assertEquals(64, parsed?.name?.length)
    }

    @Test
    fun `parsePairUri bozuk yuzde kodlamasinda patlamaz`() {
        val parsed = Pairing.parsePairUri("horus://pair?code=H7K29M4P&host=1.2.3.4&port=8088&name=%ZZ")
        assertNotNull("bozuk kodlama URI'yi tamamen reddetmemeli", parsed)
    }

    @Test
    fun `parsePairUri yinelenen anahtarda ilkini alir`() {
        val parsed = Pairing.parsePairUri(
            "horus://pair?code=H7K29M4P&host=1.2.3.4&port=8088&host=9.9.9.9",
        )
        assertEquals("1.2.3.4", parsed?.host)
    }

    @Test
    fun `parsePairUri sonucu dogrudan ayarlara yazilabilir`() {
        // Sözleşme: baseUrl mevcut istemcinin beklediği biçimde olmalı.
        val parsed = Pairing.parsePairUri("horus://pair?code=H7K29M4P&host=192.168.1.20&port=8088")
        requireNotNull(parsed)
        assertTrue(parsed.baseUrl.endsWith("/v1"))
        assertTrue(parsed.baseUrl.startsWith("http"))
    }
}

File: gateway/test/pairing.test.mjs
import { test } from "node:test";
import assert from "node:assert/strict";
import {
  PAIRING_ALPHABET,
  PAIRING_CODE_LENGTH,
  newPairingCode,
  normalizePairingCode,
  formatPairingCode,
  pairBaseUrl,
  buildPairUri,
  parsePairUri,
} from "../lib/pairing.mjs";

test("alfabede görsel ikizler yok (I, L, O, U)", () => {
  for (const ch of "ILOU") assert.equal(PAIRING_ALPHABET.includes(ch), false, ch + " alfabede olmamalı");
  assert.equal(PAIRING_ALPHABET.length, 32);
  assert.equal(new Set(PAIRING_ALPHABET).size, 32, "alfabede yinelenen karakter olmamalı");
});

test("newPairingCode doğru uzunlukta ve yalnız alfabeden üretir", () => {
  for (let i = 0; i < 200; i++) {
    const code = newPairingCode();
    assert.equal(code.length, PAIRING_CODE_LENGTH);
    for (const ch of code) assert.ok(PAIRING_ALPHABET.includes(ch), "beklenmeyen karakter: " + ch);
  }
});

test("newPairingCode enjekte edilen rastgeleliği kullanır (deterministik test)", () => {
  const code = newPairingCode(() => 0);
  assert.equal(code, "00000000");
});

test("normalizePairingCode görsel ikizleri düzeltir", () => {
  assert.equal(normalizePairingCode("h7k2-9m4p"), "H7K29M4P");
  assert.equal(normalizePairingCode("O0IL1234"), "00111234", "O→0, I→1, L→1");
  assert.equal(normalizePairingCode("  H7K2 9M4P  "), "H7K29M4P");
  assert.equal(normalizePairingCode("H7K2_9M4P"), "H7K29M4P");
});

test("normalizePairingCode geçersiz girdide kısmi kod uydurmaz", () => {
  assert.equal(normalizePairingCode(""), "");
  assert.equal(normalizePairingCode(null), "");
  assert.equal(normalizePairingCode("H7K2"), "", "kısa");
  assert.equal(normalizePairingCode("H7K29M4PX"), "", "uzun");
  assert.equal(normalizePairingCode("H7K29M4U"), "", "U alfabede yok");
  assert.equal(normalizePairingCode("H7K29M4!"), "", "alfabe dışı");
});

test("formatPairingCode insan gözü için ayırır, geçersizde girdiyi bozmaz", () => {
  assert.equal(formatPairingCode("h7k29m4p"), "H7K2-9M4P");
  assert.equal(formatPairingCode("bozuk"), "bozuk");
});

test("pairBaseUrl her zaman /v1 ile biter", () => {
  assert.equal(pairBaseUrl({ host: "192.168.1.20", port: 8088 }), "http://192.168.1.20:8088/v1");
  assert.equal(pairBaseUrl({ host: "horus.local", port: 443, tls: true }), "https://horus.local/v1");
  assert.equal(pairBaseUrl({ host: "horus.local", port: 80 }), "http://horus.local/v1");
  assert.equal(pairBaseUrl({ host: "fe80::1", port: 8088 }), "http://[fe80::1]:8088/v1", "IPv6 köşeli paranteze alınır");
});

test("pairBaseUrl geçersiz girdide boş döner", () => {
  assert.equal(pairBaseUrl({ host: "", port: 8088 }), "");
  assert.equal(pairBaseUrl({ host: "1.2.3.4", port: 0 }), "");
  assert.equal(pairBaseUrl({ host: "1.2.3.4", port: 70000 }), "");
  assert.equal(pairBaseUrl({ host: "1.2.3.4", port: "abc" }), "");
});

test("buildPairUri → parsePairUri gidiş-dönüş", () => {
  const uri = buildPairUri({ host: "192.168.1.20", port: 8088, code: "h7k2-9m4p", name: "SALIH-PC" });
  assert.ok(uri.startsWith("horus://pair?"));
  const parsed = parsePairUri(uri);
  assert.deepEqual(parsed, {
    version: 1,
    code: "H7K29M4P",
    host: "192.168.1.20",
    port: 8088,
    tls: false,
    baseUrl: "http://192.168.1.20:8088/v1",
    name: "SALIH-PC",
  });
});

test("buildPairUri tls bayrağını taşır", () => {
  const uri = buildPairUri({ host: "horus.example", port: 443, code: "H7K29M4P", tls: true });
  assert.equal(parsePairUri(uri).baseUrl, "https://horus.example/v1");
  assert.equal(parsePairUri(uri).tls, true);
});

test("buildPairUri bozuk girdide QR üretmez", () => {
  assert.equal(buildPairUri({ host: "1.2.3.4", port: 8088, code: "kısa" }), "");
  assert.equal(buildPairUri({ host: "", port: 8088, code: "H7K29M4P" }), "");
});

test("parsePairUri tanınmayan girdide null döner", () => {
  assert.equal(parsePairUri(""), null);
  assert.equal(parsePairUri(null), null);
  assert.equal(parsePairUri("https://example.com"), null, "başka şema");
  assert.equal(parsePairUri("horus://other?code=H7K29M4P"), null, "başka eylem");
  assert.equal(parsePairUri("horus://pair?v=2&code=H7K29M4P&host=1.2.3.4&port=1"), null, "bilinmeyen sürüm");
  assert.equal(parsePairUri("horus://pair?code=H7K29M4P"), null, "host/port yok");
  assert.equal(parsePairUri("horus://pair?host=1.2.3.4&port=8088"), null, "kod yok");
});

test("parsePairUri v parametresi yoksa 1 varsayar", () => {
  const parsed = parsePairUri("horus://pair?code=H7K29M4P&host=1.2.3.4&port=8088");
  assert.equal(parsed.version, 1);
  assert.equal(parsed.baseUrl, "http://1.2.3.4:8088/v1");
});

test("parsePairUri isim alanını sınırlar", () => {
  const long = "A".repeat(200);
  const parsed = parsePairUri(buildPairUri({ host: "1.2.3.4", port: 8088, code: "H7K29M4P", name: long }));
  assert.equal(parsed.name.length, 64);
});
